tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/TypedGradientAdapter.java

```diff
@@ -0,0 +1,88 @@
+/*
+ Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================
+*/
+package org.tensorflow.op;
+
+import static org.tensorflow.internal.c_api.global.tensorflow.StatusFromTF_Status;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
+import org.bytedeco.javacpp.PointerScope;
+import org.tensorflow.GradientAdapterHelpers;
+import org.tensorflow.Graph;
+import org.tensorflow.Operand;
+import org.tensorflow.Output;
+import org.tensorflow.internal.c_api.GradFunc;
+import org.tensorflow.internal.c_api.NativeOperation;
+import org.tensorflow.internal.c_api.NativeOutputVector;
+import org.tensorflow.internal.c_api.NativeStatus;
+import org.tensorflow.internal.c_api.TF_Scope;
+import org.tensorflow.internal.c_api.TF_Status;
+
+/** A native adapter for {@link CustomGradient}. */
+public final class TypedGradientAdapter<T extends RawOp> extends GradFunc {
+
+  private final CustomGradient<T> gradient;
+  private final Class<T> opClass;
+  private final Constructor<T> ctor;
+
+  public TypedGradientAdapter(CustomGradient<T> gradient, Class<T> opClass) {
+    super();
+    this.gradient = gradient;
+    this.opClass = opClass;
+    //noinspection unchecked
+    this.ctor = (Constructor<T>) opClass.getDeclaredConstructors()[0];
+    ctor.setAccessible(true);
```

Comment:
Not sure this operation will pass all security controls... looks like we are missing an easy way to convert an `Operation` to a op instance. If we keep this implementation, I would mark it as temporary and experimental until we add something better.

Reply:
Yeah, it's not great. One alternative is requiring the constructor as part of the register method, but I don't like that either. Maybe I could do something as part of the codegen, but I'm not sure what. Ideas welcome.

Reply:
@JimClarke5 , didn't you looked at another case where you wanted to get back an op class from an existing graph instead of a raw `GraphOperation`? Does that ring a bell to you?

Reply:
@JimClarke5 
> Down deep I have to port the c gradient functions to Java

Keep in mind the `tensorflow/cc/gradients` are the ones that Graph already has, so don't port those.  Hopefully we'll get the new API soon0sh anyways, so I wouldn't invest too much in this.

@karllessard 
The only information GraphOperation has about the op class is the name, so going from GraphOperation -> Op class is very hard.  I'm going from OpClass -> name, which easier.  I think rather than checking the static field, I'll just generate an annotation.  That will work for now, later on we can look at generating something like `TensorTypeRegistry`.

Reply:
Actually, that annotation will solve a few other problems of mine, too.

Reply:
What do you think about making the generated constructors public (or having a similar static factory)?  It's necessary for anyone doing Operation -> Op, and means I don't have to use `setAccessible`.